### ISSUE TITLE
Proposed contribution tasks for Outreachy applicants: Register UCSC genome canFam6

### ISSUE THREAD

**Issue by hpages** [Bioconductor/GenomeInfoDb#43]
`canFam6` is the latest UCSC genome for Dog (_Canis lupus familiaris_). See "List of UCSC genome releases" at https://genome.ucsc.edu/FAQ/FAQreleases.html for all the genomes currently supported by UCSC.

Also check out the "Genome Browser Gateway" page [here](https://genome.ucsc.edu/cgi-bin/hgGateway). This is the main entrance to the "UCSC Genome Browser". Find Dog in the UCSC species tree on the left, click on it, then make sure to select the latest Dog Assembly (`canFam6`). This will display a bunch of additional information about the `canFam6` assembly.

Note that many UCSC genomes are already registered in the **GenomeInfoDb** package (83 as of October 2022). The `registered_UCSC_genomes()` function in **GenomeInfoDb** returns the list of all the UCSC genomes that are currently registered in the package. An important thing to be aware of is that `getChromInfoFromUCSC()` still works on an unregistered genome, but in "degraded" mode, that is:
- the `assembled.molecules` argument is ignored,
- the `assembled` and `circular` columns of the returned data.frame are filled with `NA`s,
- and the chromosomes/sequences are not returned in any particular order.

Registering a genome fixes that. In other words, once a genome is registered in **GenomeInfoDb**, the information returned by `getChromInfoFromUCSC()` for that genome is guaranteed to be complete and accurate. 

See `?getChromInfoFromUCSC` (after loading **GenomeInfoDb**) for more information.

Registering a new UCSC genome is only a matter of adding a new file, called "registration file", to [`GenomeInfoDb/inst/registered/UCSC_genomes/`](https://github.com/Bioconductor/GenomeInfoDb/tree/master/inst/registered/UCSC_genomes). Note that the folder contains a `README.TXT` file that provides some brief information about what a "registration file" should contain (unfortunately the registration process is not fully documented).

For `canFam6`, the easiest way to go would be to copy `canFam5.R` -> `canFam6.R`, then to make some adjustments to `canFam6.R`. Don't bother with the `NCBI_LINKER` component for now. We'll add it later, once the corresponding NCBI assembly (`Dog10K_Boxer_Tasha`) is also registered (registering `Dog10K_Boxer_Tasha` is the topic of issue #44).

_IMPORTANT NOTES TO OUTREACHY APPLICANTS:_
- _Make sure to complete all the **Preliminary tasks** [listed here](https://github.com/Bioconductor/BSgenomeForge/wiki/List-of-contribution-tasks-for-the-Outreachy-application-period) before you start working on this task. In particular, make sure that you have R 4.2 and that you are set up to use the **devel** version of Bioconductor (currently 3.16)._
- _Only one applicant can work on this task. If you choose to work on this task, please make sure to assign yourself so other applicants know that the task is already being worked on. If later on you change your mind, please unassign yourself. It's ok to change your mind!_
- _To work on this task, please fork the **GenomeInfoDb** repository. Then do your work on that fork._
- _Always test your changes before you commit them to your fork. This consists in installing the modified package, starting R, loading the package, and playing around with the new functionality. This process is called "ad hoc manual testing". Once everything behaves and looks as expected, run `R CMD build` and `R CMD check` on the package. Note that `R CMD check` should always be run on the **source tarball** produced by `R CMD build`._
- _`R CMD check` might produce some NOTEs and even some WARNINGs. These are ok if they existed before your changes. You can check that by taking a look at the daily report produced by our automated builds here: https://bioconductor.org/checkResults/devel/bioc-LATEST/ Make sure to not introduce new NOTEs or WARNINGs!_
- _Once your work is ready to be merged, please submit a PR (Pull Request)._
- _Remember to record your contribution on Outreachy at https://www.outreachy.org/outreachy-december-2022-internship-round/communities/bioconductor/refactor-the-bsgenomeforge-tools/contributions/._

**Comment by hpages** [comment]
Hi @Priceless-P ,

Welcome to Outreachy and thanks for volunteering for this task!

Please confirm again that you have successfully completed all the Preliminary tasks listed [here](https://github.com/Bioconductor/BSgenomeForge/wiki/List-of-contribution-tasks-for-the-Outreachy-application-period). Do you have any question about these tasks? Don't hesitate to ask. You can ask me by email or in the #outreachy channel on the community-bioc Slack (don't ask questions about these Preliminary tasks here, in this issue, to stay on-topic).

To assign yourself to this task, click on the "assign yourself" link at the top of the right margin of this issue.

Note that I've just committed a few changes to **GenomeInfoDb** this morning. So if you have already forked this repository, you'd need to update your fork.

Thanks,
H.

**Comment by Priceless-P** [comment]
@hpages Yes, I have completed all the preliminary tasks stated. I didn’t encounter any issues with them. I will go ahead and assign myself now. I will also update my fork and start working on it.

**Comment by Priceless-P** [comment]
@hpages , I can't see the button to assign myself 

![IMG_2477](https://user-images.githubusercontent.com/88077456/195689318-54a16d6e-a19b-4d85-b7c8-f8fb7c47d750.jpeg)

**Comment by hpages** [comment]
Oh sorry! Probably because you don't have the permissions to do that. I just did it for you.

**Comment by hpages** [comment]
@Priceless-P Before running `R CMD build` or `R CMD check`, you want to make sure that you also have all the packages listed in the `Suggests` field. Packages listed in this field are typically used in the vignettes or in some examples, so they are not strictly needed to _use_ the package, but they are needed to build the vignettes or to run the examples in the package. That's why `R CMD build` and `R CMD check` need them.

There are several ways to check that you have all the packages needed for running `R CMD build` and `R CMD check`. The way I usually deal with this is:

1. Use the `--no-build-vignettes` flag when running `R CMD build` on the package. This will quickly produce the package source tarball. It should always work, whether you have all the dependencies installed or not. Note that the resulting tarball is incomplete because it's missing the vignettes but it doesn't matter for now.
2. Run `R CMD check` on the incomplete tarball. `R CMD check` will fail early if you don't have all the necessary packages. The error message will tell you which one are missing.
3. Use `BiocManager::install()` to install the missing packages.
4. Now try to run `R CMD build` again. This time it will take longer because it will build the vignettes. It might still fail, but not because you have missing deps. If successful, the resulting tarball will be complete i.e. it will contain the vignettes.
5. Run `R CMD check` again on the complete tarball.

Other developers might use a different way. For example I learned recently that if you install a package with `BiocManager::install(.. , dependencies=TRUE)`, this will also take care of installing the packages listed in the `Suggests` field. So this might actually be easier than what I'm doing!

Note that the name of your registration file should be `canFam6.R`, not `cFam6.R`.

Let me know how it goes.

**Comment by Priceless-P** [comment]
Okay. I'll do it your way. 

Yes, `cFam6.R` i misspelled that.

**Comment by Priceless-P** [comment]
After building the second time (with vignettes) as you stated in point 4 above

I got the following errors
```
Error in stop_if(is.null(NCBI_assembly_info), "\"assembly_accession\" field in 'NCBI_LINKER' must ",  : 
  Error in UCSC genome registration file 'canFam6.R':
  "assembly_accession" field in 'NCBI_LINKER' must be associated with a
  registered NCBI assembly
Calls: registered_UCSC_genomes -> lapply -> lapply -> FUN -> stop_if
Execution halted
* checking for unstated dependencies in ‘tests’ ... OK
* checking tests ...
  Running ‘run_unitTests.R’
 ERROR
Running the tests in ‘tests/run_unitTests.R’ failed.
Last 13 lines of output:
  ERROR in test_seqlevelsStyle_Seqinfo: Error in stop_if(is.null(NCBI_assembly_info), "\"assembly_accession\" field in 'NCBI_LINKER' must ",  : 
    Error in UCSC genome registration file 'canFam6.R':
    "assembly_accession" field in 'NCBI_LINKER' must be associated with a
    registered NCBI assembly
  
  Test files with failing tests
  
     test_seqlevelsStyle.R 
       test_seqlevelsStyle_Seqinfo 
  
  Error in BiocGenerics:::testPackage("GenomeInfoDb") : 
    unit tests failed for package GenomeInfoDb
  Calls: <Anonymous> -> <Anonymous>
  Execution halted
  [......]
* DONE

Status: 2 ERRORs, 2 NOTEs
```

I used the command `R CMD check GenomeInfoDb_1.33.9.tar.gz `

But I noticed that [here](https://bioconductor.org/checkResults/devel/bioc-LATEST/GenomeInfoDb/lconway-checksrc.html) they used this command

`
 /Library/Frameworks/R.framework/Resources/bin/R CMD check --install=check:GenomeInfoDb.install-out.txt --library=/Library/Frameworks/R.framework/Resources/library --no-vignettes --timings GenomeInfoDb_1.33.8.tar.gz
`

You said to ignore the `NCBI_LINKER` for now, but what about the second one? Is it something I did wrong?

**Comment by hpages** [comment]
`NCBI_LINKER` is optional. When I said to ignore it, I meant to not put it in the registration file. You can not "link" the `canFam6` genome to its NCBI assembly (`Dog10K_Boxer_Tasha`) at the moment, because the latter is not registered yet.

> I used the command R CMD check GenomeInfoDb_1.33.9.tar.gz

Yep, that's the right command. Our build system uses a more complicated command, but complicated reasons that I won't discuss here.

Before you run `R CMD build` and `R CMD check`, make sure to spend some time testing the new functionality in an interactive session. At the very least, you should try to call `getChromInfoFromUCSC("canFam6")`. Try to call it in different ways e.g. with and without `assembled.molecules.only` set to TRUE, with and without `map.NCBI` set to TRUE, etc... and see how the function behaves. This is what I call ad hoc manual testing. Only when everything looks ok should you run `R CMD build` and `R CMD check`.

**Comment by Priceless-P** [comment]
Okay, Will do.
Thanks!

**Comment by Priceless-P** [comment]
@hpages I ran into another error when i executed `FETCH_ORDERED_CHROM_SIZES()`

```
Error in .order_seqlevels(chrom_sizes[, "chrom"]) : 
all(npart <= 2L) is not TRUE
```

This is what i currently have 

```
[.....]
ASSEMBLED_MOLECULES <- paste0("chr", c(1, "X", 3:38, "M"))
CIRC_SEQS <- "chrM"

[...]
.order_seqlevels <- function(seqlevels)
{
  tmp <- CharacterList(strsplit(seqlevels, "_"))
  npart <- lengths(tmp)
  stopifnot(all(npart <= 2L))
  
  idx1 <- which(npart == 1L)
  stopifnot(length(idx1) == length(ASSEMBLED_MOLECULES))
  oo1 <- match(ASSEMBLED_MOLECULES, seqlevels[idx1])
  stopifnot(!anyNA(oo1))
  idx1 <- idx1[oo1]
  
  idx2 <- which(npart == 2L)
  m2 <- matrix(unlist(tmp[idx2]), ncol=2L, byrow=TRUE)
  stopifnot(all(m2[ , 1L] == "chrUn"))
  oo2 <- order(m2[ , 2L])
  idx2 <- idx2[oo2]
  
  c(idx1, idx2)
}

FETCH_ORDERED_CHROM_SIZES <-
  function(goldenPath.url=getOption("UCSC.goldenPath.url"))
  {
    chrom_sizes <- GenomeInfoDb:::fetch_chrom_sizes_from_UCSC(GENOME,
                                                              goldenPath.url=goldenPath.url)
    oo <- .order_seqlevels(chrom_sizes[ , "chrom"])
    S4Vectors:::extract_data_frame_rows(chrom_sizes, oo)
  }
```

Is my `ASSEMBLED_MOLECULES` wrong? Please how do i go about this?

**Comment by hpages** [comment]
First of all, we want to list the assembled molecules in the traditional order: numbered chromosomes first, then sex chromosome, then mitochondrial chromosome. Also there's no reason to exclude chromosome 2. So:
```
ASSEMBLED_MOLECULES <- paste0("chr", c(1:38, "X", "M"))
```
Note that this is what all other `canFam*.R` files have.

Now let's take a closer look at all the sequences in `canFam6`:
```
> library(GenomeInfoDb)
> chrominfo <- getChromInfoFromUCSC("canFam6")
> chrominfo$chrom
  [1] "chr1"                 "chr2"                 "chr3"                
  [4] "chr4"                 "chr5"                 "chr6"                
  [7] "chr7"                 "chr8"                 "chr9"                
 [10] "chr10"                "chr11"                "chr12"               
 [13] "chr13"                "chr14"                "chr15"               
 [16] "chr16"                "chr17"                "chr18"               
 [19] "chr19"                "chr20"                "chr21"               
 [22] "chr22"                "chr23"                "chr24"               
 [25] "chr25"                "chr26"                "chr27"               
 [28] "chr28"                "chr29"                "chr30"               
 [31] "chr31"                "chr32"                "chr33"               
 [34] "chr34"                "chr35"                "chr36"               
 [37] "chr37"                "chr38"                "chrX"                
 [40] "chrM"                 "chrUn_NW_023329667v1" "chrUn_NW_023329668v1"
 [43] "chrUn_NW_023329669v1" "chrUn_NW_023329670v1" "chrUn_NW_023329671v1"
 [46] "chrUn_NW_023329672v1" "chrUn_NW_023329673v1" "chrUn_NW_023329674v1"
 [49] "chrUn_NW_023329675v1" "chrUn_NW_023329676v1" "chrUn_NW_023329677v1"
...
[139] "chrUn_NW_023329765v1" "chrUn_NW_023329766v1" "chrUn_NW_023329767v1"
[142] "chrUn_NW_023329768v1" "chrUn_NW_023329769v1" "chrUn_NW_023329770v1"
[145] "chrUn_NW_023329771v1" "chrUn_NW_023329772v1" "chrUn_NW_023329773v1"
```
All the additional sequences are called "unplaced scaffolds". Those are small DNA fragments that were produced by the assembly process but the scientists working on the assembly were not able to identify which chromosome those fragments are coming from. UCSC naming convention is to prefix the names of  "unplaced scaffolds" with `chrUn_`.

Note that the set of unplaced scaffolds was very different in the previous version of the assembly:
```
> chrominfo <- getChromInfoFromUCSC("canFam5")
> chrominfo$chrom
  [1] "chr1"                 "chr2"                 "chr3"                
  [4] "chr4"                 "chr5"                 "chr6"                
  [7] "chr7"                 "chr8"                 "chr9"                
 [10] "chr10"                "chr11"                "chr12"               
 [13] "chr13"                "chr14"                "chr15"               
 [16] "chr16"                "chr17"                "chr18"               
 [19] "chr19"                "chr20"                "chr21"               
 [22] "chr22"                "chr23"                "chr24"               
 [25] "chr25"                "chr26"                "chr27"               
 [28] "chr28"                "chr29"                "chr30"               
 [31] "chr31"                "chr32"                "chr33"               
 [34] "chr34"                "chr35"                "chr36"               
 [37] "chr37"                "chr38"                "chrX"                
 [40] "chrM"                 "chrUn_REHQ01000041v1" "chrUn_REHQ01000042v1"
 [43] "chrUn_REHQ01000043v1" "chrUn_REHQ01000044v1" "chrUn_REHQ01000045v1"
 [46] "chrUn_REHQ01000046v1" "chrUn_REHQ01000047v1" "chrUn_REHQ01000048v1"
 [49] "chrUn_REHQ01000049v1" "chrUn_REHQ01000050v1" "chrUn_REHQ01000051v1"
 ...
[787] "chrUn_REHQ01000787v1" "chrUn_REHQ01000788v1" "chrUn_REHQ01000789v1"
[790] "chrUn_REHQ01000790v1" "chrUn_REHQ01000791v1" "chrUn_REHQ01000792v1"
[793] "chrUn_REHQ01000793v1" "chrUn_REHQ01000794v1"
```
Even though the "unplaced scaffolds" are still prefixed with `chrUn_` (as they should), what comes after the prefix is very different. One important difference is that the names have 2 parts in `canFam5`, but now in `canFam6` they have 3 parts.

These lines in `canFam5.R`:
```
tmp <- CharacterList(strsplit(seqlevels, "_"))    # 1.
npart <- lengths(tmp)                             # 2.
stopifnot(all(npart <= 2L))                       # 3.
```
are doing the following:
1. Break the `seqlevels` (a.k.a. sequence names) in parts. The parts are returned in a CharacterList object.
2. Compute the number of parts in each seqlevel. The result (`npart`) is an integer vector _parallel_ to `tmp` (i.e. same length, and the i-th element in `npart` corresponds to the i-th element in `tmp`).
3. Check that all seqlevels have at most 2 parts. This is just a sanity check.

Please take the time to run the code line by line and to look at what it does. Inspect all intermediate results. You're going to dive deep into the guts of R. There's a lot going on with those 3 lines of code so it's going to be a great learning experience.

For `canFam6`, the number of parts are either 1 (for chromosomes) or 3 (for unplaced scaffolds).

If you open a terminal and go to the `GenomeInfoDb/inst/registered/UCSC_genomes/` folder, you can do:
```
grep stopifnot * | grep npart
```
to see what the other registration files do for this sanity check. You will see that some of them do:
```
stopifnot(all(npart %in% c(1L, 3L)))
```
This means that the sanity check is checking that all the seqlevels have either 1 or 3 parts.

Take a look at those files to see how their `.order_seqlevels()` function handles things. A good example is `bosTau9.R`.

Hope this helps,
H.

**Comment by Priceless-P** [comment]
Yes. It really does. Thanks a lot @hpages You’re the best! Please check your slack. I have just one last question.

**Comment by hpages** [comment]
I just merged PR #52. Congratulation @Priceless-P on your first contribution to Bioconductor! Don't forget to record it on Outreachy at https://www.outreachy.org/outreachy-december-2022-internship-round/communities/bioconductor/refactor-the-bsgenomeforge-tools/contributions/.

The next suggested task for you is #44. Whenever you are ready. Thanks!
